### PR TITLE
Do not run 4GB string tests on low memory systems (fixes #38392)

### PR DESCRIPTION
**Pull request by xmcclure**
"Low memory" here is defined as 32-bit, less than 6GB of physical RAM, or mobile (since mobile lacks the class for querying physical RAM).

Without this patch, one of the corlib tests allocates 4 GB of RAM, which causes OOM crashes on low-memory Jenkins instances and also mobile phones.

### CODE REVIEW

File: mcs/class/corlib/Test/System/StringTest.cs [hunk 1]
```diff
@@ -92,19 +96,36 @@ public void Constructor4_Count_Negative ()
...
+#if MOBILE
+		return true;
+#else
+		if (!Environment.Is64BitProcess)
+			return true;
+
+		// Require 6 GB physical RAM, for the 4GB string plus 2GB headroom 
+		var pc = new PerformanceCounter ("Mono Memory", "Total Physical Memory");
+
+		return pc.RawValue < 6L*1024L*1024L*1024L;
```
Comment (akoeplinger) [inline mcs/class/corlib/Test/System/StringTest.cs]: nit: no need for the L everywhere: `pc.RawValue < 6L * 1024 * 1024 * 1024`

File: mcs/class/corlib/Test/System/StringTest.cs [hunk 2]
```diff
@@ -92,19 +96,36 @@ public void Constructor4_Count_Negative ()
...
-			if (Environment.Is64BitProcess) {
-				Assert.AreEqual ('A', x[0]);
-				Assert.AreEqual ('A', x[int.MaxValue - 1]);
-			}
-			else
-				Assert.Fail ("Expected OutOfMemoryException.");
-		} catch (OutOfMemoryException) {
-		}
+		if (LowMemoryTestEnvironment())
+			return;
```
Comment (akoeplinger) [inline mcs/class/corlib/Test/System/StringTest.cs]: Do Assert.Ignore ("Low memory device, skipping.") here and in the other tests so we don't get a "passed" test result even though it didn't run.

**Comment by akoeplinger** [comment]
LGTM apart from the two comments. Thanks for fixing this!

File: mcs/class/corlib/Test/System/StringTest.cs (hunk 1 above)
Comment (xmcclure) [inline mcs/class/corlib/Test/System/StringTest.cs]: I'm gonna leave this as if if that's ok, the Ls are harmless and for me they save the 0.25s of valuable time trying to remember the int promotion rules when scanning the code.

File: mcs/class/corlib/Test/System/StringTest.cs (hunk 2 above)
Comment (xmcclure) [inline mcs/class/corlib/Test/System/StringTest.cs]: Hm. If I’m Assert.Ignoring on MOBILE, maybe I could remove the MobileNotWorking category.

File: mcs/class/corlib/Test/System/StringTest.cs (hunk 1 above)
Comment (akoeplinger) [inline mcs/class/corlib/Test/System/StringTest.cs]: sure

Comment (marek-safar) [inline mcs/class/corlib/Test/System/StringTest.cs]: If you are not sure then do
const long requiredMemory = 6L * ...... instead

**Comment by xmcclure** [comment]
@monojenkins merge

**Pull request merged**